sycl/doc/extensions/SubGroupAlgorithms/SYCL_INTEL_sub_group_algorithms.asciidoc

```diff
@@ -0,0 +1,177 @@
+
+All functions are supported for the fundamental scalar types supported by SYCL and instances of the SYCL +vec+ class. The fundamental scalar types (as defined in Section 6.5 of the SYCL 1.2.1 specification) are: +bool+, +char+, +signed char+, +unsigned char+, +short int+, +unsigned short int+, +int+, +unsigned int+, +long int+, +unsigned long int+, +long long int+, +unsigned long long int+, +size_t+, +float+, +double+, +half+.
+
+Functions with arguments of type +vec<T,N>+ are applied component-wise: they are semantically equivalent to N calls to a scalar function of type +T+.
+
+=== Functions
+
+The sub-group algorithms library is based on the algorithms library described in Section 28 of the {cpp}17 standard.  The syntax and restrictions are aligned, with two notable differences: the first argument to each function is a sub-group of work-items, in place of an execution policy; and pointers are accepted in place of iterators in order to guarantee that address space information is visible to the compiler.
+
+Sub-group algorithms are performed collaboratively by the work-items in a sub-group. All functions therefore act as synchronization points and must be encountered in converged control flow by all work-items in the sub-group -- if one work-item in the sub-group reaches the function, then all work-items in the sub-group must reach the function.  Additionally, restrictions may be placed on the arguments passed to each function in order to ensure that all work-items in the sub-group agree on the operation that is being performed.  Any such restrictions on the arguments passed to a function are defined within the descriptions of those functions.
+
+Many functions provide at least two overloads: one operating directly on data produced by the work-items in the sub-group, and another operating on a range of data in memory specified by a pair of pointers.  If the pointers passed to such a sub-group function are not the same for all work-items in the sub-group, their behavior is undefined.  How the elements of a range are processed by the work-items in a sub-group is undefined.
+
+It is undefined behavior for any of these functions to be invoked within a +parallel_for_work_group+ or +parallel_for_work_item+ context, but this restriction may be lifted in a future version of the proposal.
+
+==== Votes and Collectives
+
+Each of these functions from +SYCL_INTEL_group_library+ is supported for sub-groups.  The definitions are identical, except an instance of the +sub_group+ class is accepted as the first argument in place of an instance of the +group+ class.
+
+==== Permutes and Shifts
+
+The permute sub-group functions perform arbitrary communication between pairs of work-items in a sub-group.  Common patterns -- such as shifting all values in a sub-group by a fixed number of work-items, or reversing the order of all values in a sub-group -- are exposed as specialized functions that may be accelerated in hardware.
+
+|===
+|Function|Description
```

Comment:
Previous version of the spec had equivalents of  the following built-in functions from `cl_intel_subgroups` OpenCL extension: `intel_sub_group_shuffle_down` and `intel_sub_group_shuffle_up`, while this algorithms library doesn't describe them.

I see that `shift_left` and `shift_right` are the most closer ones, but unlike the OpenCL extension, `delta` has to be the same across all work-items in a sub-group, which basically disallows cycle shifts. I don't know how important that use-case is, but this functions will *almost*, I guess, result in undefined elements in some elements of sub-group

Reply:
This is a deliberate change.  Other architectures can only implement `shuffle_down`/`shuffle_up` more efficiently than a general shuffle if it's known to be a shift (i.e. that `delta` is uniform).  Developers can mimic the original behavior by using `permute` where the indices are computed from offsets.

It hadn't occurred to me that developers might be using non-uniform `delta` to implement cycle shifts.  If that use-case is important, we should consider adding an equivalent of [`std::rotate`](https://en.cppreference.com/w/cpp/algorithm/rotate).

Reply:
> It hadn't occurred to me that developers might be using non-uniform delta to implement cycle shifts. If that use-case is important, we should consider adding an equivalent of std::rotate.

I don't have any data about amount of such usages, but I guess this shouldn't be a blocker for this PR - we can always add such if requested